Windows MariaDB 11.4.5 upgrade.

TheKnowles opened the pull request:
Adds Windows MariaDB 11.4.5 binaries.

vorburger commented on DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/pom.xml:
```diff
@@ -0,0 +1,55 @@
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-antrun-plugin</artifactId>
+        <version>1.8</version>
```
```suggestion
        <version>3.1.0</version>
```

@TheKnowles I bumped this in the Linux (and now macOS as well!) ones - do you want to do that here as well?

TheKnowles replied:
Agree. Old version from copy paste from 10.6.x winx pom.

vorburger commented on DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/prepare.xml:
```diff
@@ -0,0 +1,61 @@
+      dest="${project.build.directory}/generated-resources/ch/vorburger/mariadb4j/mariadb-${project.version}/winx64"
+      overwrite="true">
+      <patternset>
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mysql.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mysqld.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mysqlcheck.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mysqldump.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mysql_install_db.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadbd.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb-check.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb-dump.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb-install-db.exe" />
+        <include name="mariadb-${mariaDB.version}-winx64/bin/my_print_defaults.exe" />
```
@TheKnowles to limit the size here, and [align it with this](https://github.com/MariaDB4j/MariaDB4j/blob/87f93cb927f77540a1672fe647dd887dfb308342/DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-11.4.5-linux-x86_64/prepare.xml#L60) [which, as far as I can tell, seems to suffice?], how would you feel about cutting it down to this, and remove both all the `mysql*` executables (old, we don't need them; anymore), as well as `mariadb-install-db` (which I don't remember exactly what it is, but it seems to work without):

```suggestion
        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb.exe" />
        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadbd.exe" />
        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb-check.exe" />
        <include name="mariadb-${mariaDB.version}-winx64/bin/mariadb-dump.exe" />
        <include name="mariadb-${mariaDB.version}-winx64/bin/my_print_defaults.exe" />
```

What's however missing here is that `resolveip` thing - I'm guessing Windows doesn't need that? Just double checking!

TheKnowles replied:
Investigating. I vaguely recall with 10.6.x I had to keep them all because Windows did not fully switch over.

TheKnowles replied:
Yes, we need to keep all of this as is. Windows mixes and matches its various install scripts with binary names. If I remove the mysql* items, the regular build process/tests will fail

```
Caused by: ch.vorburger.exec.ManagedProcessException: Program [C:\Users\katch\AppData\Local\Temp\MariaDB4j\base\bin\mariadb-install-db.exe, --datadir=C:\Users\katch\AppData\Local\Temp\MariaDB4j\data\63182] (in working directory C:\Users\katch\AppData\Local\Temp\MariaDB4j\base) failed with Exception: , last 100 lines of console:
Running bootstrap
'"C:\Users\katch\AppData\Local\Temp\MariaDB4j\base\bin\mysqld.exe"' is not recognized as an internal or external command,
operable program or batch file.
FATAL ERROR: database creation failed
FATAL ERROR: database creation failedCreating my.ini file
```

vorburger replied:
Interesting! Thank You for having had a closer look, and elaborating What a mess that is... I was curious about sizes, FTR:

```sh
$ ll DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/target/generated-resources/ch/vorburger/mariadb4j/mariadb-11.4.5/winx64/bin/
.rw-r--r--. vorburger vorburger 4.5 MB 2025-01-30 19:12  mariadb-check.exe
.rw-r--r--. vorburger vorburger 4.6 MB 2025-01-30 19:12  mariadb-dump.exe
.rw-r--r--. vorburger vorburger 3.1 MB 2025-01-30 19:12  mariadb-install-db.exe
.rw-r--r--. vorburger vorburger 4.6 MB 2025-01-30 19:12  mariadb.exe
.rw-r--r--. vorburger vorburger  24 KB 2025-01-30 19:11  mariadbd.exe
.rw-r--r--. vorburger vorburger 698 KB 2025-01-30 19:12  my_print_defaults.exe
.rw-r--r--. vorburger vorburger 4.6 MB 2025-01-30 19:12  mysql.exe
.rw-r--r--. vorburger vorburger 3.1 MB 2025-01-30 19:12  mysql_install_db.exe
.rw-r--r--. vorburger vorburger 4.5 MB 2025-01-30 19:12  mysqlcheck.exe
.rw-r--r--. vorburger vorburger  24 KB 2025-01-30 19:11  mysqld.exe
.rw-r--r--. vorburger vorburger 4.6 MB 2025-01-30 19:12  mysqldump.exe
.rw-r--r--. vorburger vorburger  20 MB 2025-01-30 19:10  server.dll
.rw-r--r--. vorburger vorburger 8.6 MB 2025-01-30 19:10  server.lib

$ du -h DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/target/generated-resources/ch/vorburger/mariadb4j/mariadb-11.4.5/winx64/bin/
63M     DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/target/generated-resources/ch/vorburger/mariadb4j/mariadb-11.4.5/winx64/bin/
```

So the bulk (20 + 8.6 MB) is (probably shared from) `server.dll` + `server.lib`, and the "duplication" is 4.5+4.6+3.1+4.6=16.8 MB, which is ~27% of 63 MB.

From your test, it looks like `mariadb-install-db.exe` needs `mysqld.exe` (weird!) - but maybe "the other way around" is possible, and we could just keep only `mysql*.exe`? I'll have a quick go at it to try - just for "fun of" optimizing JAR sizes...

PS: Maybe I'll also report this "upstream".

vorburger replied:
> From your test, it looks like mariadb-install-db.exe needs mysqld.exe (weird!) - but maybe "the other way around" is possible, and we could just keep only mysql*.exe? I'll have a quick go at it to try - just for "fun of" optimizing JAR sizes...

I've tried this in #1134 and it worked! 🗜️

vorburger commented:
@TheKnowles awesome!! 🚀 Thank You so much for helping and jumping in here with this, for #1120.

To test if this works on CI (before I'll `mvn deploy` it to Maven Central), could you [just temporarily, I'll remove it again later: once merged and when it's on Maven Central] add (somethign like) a `./mvnw.cmd -f DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/pom.xml` to the `testOnWindows` in the .github/workflows/test.yaml - do you know what I mean?

Also, if you wouldn't mind to do a rebase now, we'll include the #1125 - which includes brand new macOS coverage!

vorburger commented:
@TheKnowles I've had a look, and there are x2 unrelated problems here now:

On Windows, it fails due to:

```
Run ankane/setup-mariadb@14ea302aa8c921b5a55450add91abb724369d40b

D:\a\_actions\ankane\setup-mariadb\14ea302aa8c921b5a55450add91abb[7](https://github.com/MariaDB4j/MariaDB4j/actions/runs/14140666048/job/39621385436?pr=1126#step:6:8)24369d40b\index.js:4[8](https://github.com/MariaDB4j/MariaDB4j/actions/runs/14140666048/job/39621385436?pr=1126#step:6:9)
  throw 'Invalid MariaDB version: ' + mariadbVersion;
  ^
Invalid MariaDB version: 11.4
(Use `node --trace-uncaught ...` to show where the exception was thrown)

Node.js v20.1[9](https://github.com/MariaDB4j/MariaDB4j/actions/runs/14140666048/job/39621385436?pr=1126#step:6:10).0
```

I haven't looked more closely into it, but that's probably something you would want to fix, before I can merge this?

On Linux and the (new) macOS it fails because that `-f DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/pom.xml` actually needs to be added to (both) of them as well, not just Windows; I did not think of that when I originally suggested this. If you could do that so that we can get a green build here, I'll merge it (and then afterwards remove all x3 again, once I've deployed it to Maven central).

TheKnowles commented:
> @TheKnowles I've had a look, and there are x2 unrelated problems here now:
> 
> On Windows, it fails due to:
> 
> ```
> Run ankane/setup-mariadb@14ea302aa8c921b5a55450add91abb724369d40b
> 
> D:\a\_actions\ankane\setup-mariadb\14ea302aa8c921b5a55450add91abb[7](https://github.com/MariaDB4j/MariaDB4j/actions/runs/14140666048/job/39621385436?pr=1126#step:6:8)24369d40b\index.js:4[8](https://github.com/MariaDB4j/MariaDB4j/actions/runs/14140666048/job/39621385436?pr=1126#step:6:9)
>   throw 'Invalid MariaDB version: ' + mariadbVersion;
>   ^
> Invalid MariaDB version: 11.4
> (Use `node --trace-uncaught ...` to show where the exception was thrown)
> 
> Node.js v20.1[9](https://github.com/MariaDB4j/MariaDB4j/actions/runs/14140666048/job/39621385436?pr=1126#step:6:10).0
> ```
> 
> I haven't looked more closely into it, but that's probably something you would want to fix, before I can merge this?
> 
> On Linux and the (new) macOS it fails because that `-f DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-winx64-11.4.5/pom.xml` actually needs to be added to (both) of them as well, not just Windows; I did not think of that when I originally suggested this. If you could do that so that we can get a green build here, I'll merge it (and then afterwards remove all x3 again, once I've deployed it to Maven central).

Rebased. All three OS tests now pass with the addition of installing the new windows 11.4.5 artifact. Added this to the CodeQL action as well.

vorburger commented:
> Rebased. All three OS tests now pass with the addition of installing the new windows 11.4.5 artifact. Added this to the CodeQL action as well.

Awesome!! 🚀 I'm now merging this, as-is ([pre-commit still fails](https://results.pre-commit.ci/run/github/3355640/1743214072.OqidjNgJTziZiuXsqnHKxg); I'll fix it just after).

vorburger commented:
FTR: I have now deployed this (with #1134) to Maven Central; see https://repo.maven.apache.org/maven2/ch/vorburger/mariaDB4j/mariaDB4j-db-winx64/.

#1136 removes the `mvn -f DB` again which we temporarily introduced here.

With that, this (and #1120) is _"full done"._ 🚀
